From edbb87149952d8c80e5303eadc3ec5b3f91e992a Mon Sep 17 00:00:00 2001
From: Wendy Liang <wendy.liang@xilinx.com>
Date: Tue, 4 Aug 2020 13:42:11 -0700
Subject: [PATCH 1500/1852] misc: xilinx-ai-engine: use bitmask for AI engine
 tile type

commit bcff829d63749e734518b4376472776952e6471f from
https://github.com/Xilinx/linux-xlnx.git

This is to fix the register type check in register verification. There
are some registers available in more than one type of tiles. Such as
registers available to SHIM PL are also available to SHIM NOC. And thus,
bits mask for AIE tile types is easier to indicate the registers are
available in which types of tiles.

This information is used to validate registers access request from users
to see if the registers should not be configured from user of a tile.

Signed-off-by: Wendy Liang <wendy.liang@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/misc/xilinx-ai-engine/ai-engine-internal.h | 13 ++++++++-----
 drivers/misc/xilinx-ai-engine/ai-engine-v1.c       | 14 +++++++++-----
 2 files changed, 17 insertions(+), 10 deletions(-)

diff --git a/drivers/misc/xilinx-ai-engine/ai-engine-internal.h b/drivers/misc/xilinx-ai-engine/ai-engine-internal.h
index 7d44d34aa82d..a85e9a9758df 100644
--- a/drivers/misc/xilinx-ai-engine/ai-engine-internal.h
+++ b/drivers/misc/xilinx-ai-engine/ai-engine-internal.h
@@ -9,6 +9,7 @@
 #define AIE_INTERNAL_H
 
 #include <linux/bitfield.h>
+#include <linux/bits.h>
 #include <linux/cdev.h>
 #include <linux/device.h>
 #include <linux/file.h>
@@ -20,11 +21,13 @@
 #include <linux/of_device.h>
 #include <uapi/linux/xlnx-ai-engine.h>
 
-enum aie_tile_type {
-	AIE_TILE_TYPE_TILE = 1,
-	AIE_TILE_TYPE_SHIMPL,
-	AIE_TILE_TYPE_SHIMNOC
-};
+/*
+ * Macros for AI engine tile type bitmasks
+ */
+#define AIE_TILE_TYPE_TILE	BIT(0)
+#define AIE_TILE_TYPE_SHIMPL	BIT(1)
+/* SHIM NOC tile includes SHIM PL and SHIM NOC modules */
+#define AIE_TILE_TYPE_SHIMNOC	BIT(2)
 
 /*
  * Macros for attribute property of AI engine registers accessed by kernel
diff --git a/drivers/misc/xilinx-ai-engine/ai-engine-v1.c b/drivers/misc/xilinx-ai-engine/ai-engine-v1.c
index 1520a9fdd98a..3defece1a2bb 100644
--- a/drivers/misc/xilinx-ai-engine/ai-engine-v1.c
+++ b/drivers/misc/xilinx-ai-engine/ai-engine-v1.c
@@ -67,27 +67,31 @@ static const struct aie_tile_regs aiev1_kernel_regs[] = {
 	 .eoff = AIE_SHIMNOC_L2INTR_INTR_REGOFF,
 	},
 	/* SHIM 1st level interrupt controller */
-	{.attribute = AIE_TILE_TYPE_SHIMPL << AIE_REGS_ATTR_TILE_TYPE_SHIFT,
+	{.attribute = (AIE_TILE_TYPE_SHIMPL | AIE_TILE_TYPE_SHIMNOC) <<
+		      AIE_REGS_ATTR_TILE_TYPE_SHIFT,
 	 .soff = AIE_SHIMPL_L1INTR_MASK_A_REGOFF,
 	 .eoff = AIE_SHIMPL_L1INTR_BLOCK_NORTH_B_REGOFF,
 	},
 	/* SHIM column reset */
-	{.attribute = AIE_TILE_TYPE_SHIMPL << AIE_REGS_ATTR_TILE_TYPE_SHIFT,
+	{.attribute = (AIE_TILE_TYPE_SHIMPL | AIE_TILE_TYPE_SHIMNOC) <<
+		      AIE_REGS_ATTR_TILE_TYPE_SHIFT,
 	 .soff = AIE_SHIMPL_COLRESET_REGOFF,
 	 .eoff = AIE_SHIMPL_COLRESET_REGOFF,
 	},
 	/* SHIM reset Enable */
-	{.attribute = AIE_TILE_TYPE_SHIMPL << AIE_REGS_ATTR_TILE_TYPE_SHIFT,
+	{.attribute = (AIE_TILE_TYPE_SHIMPL | AIE_TILE_TYPE_SHIMNOC) <<
+		      AIE_REGS_ATTR_TILE_TYPE_SHIFT,
 	 .soff = AIE_SHIMPL_RESET_REGOFF,
 	 .eoff = AIE_SHIMPL_RESET_REGOFF,
 	},
 	/* SHIM clock control */
-	{.attribute = AIE_TILE_TYPE_SHIMPL << AIE_REGS_ATTR_TILE_TYPE_SHIFT,
+	{.attribute = (AIE_TILE_TYPE_SHIMPL | AIE_TILE_TYPE_SHIMNOC) <<
+		      AIE_REGS_ATTR_TILE_TYPE_SHIFT,
 	 .soff = AIE_SHIMPL_CLKCNTR_REGOFF,
 	 .eoff = AIE_SHIMPL_CLKCNTR_REGOFF,
 	},
 	/* Tile clock control */
-	{.attribute = AIE_TILE_TYPE_SHIMPL << AIE_REGS_ATTR_TILE_TYPE_SHIFT,
+	{.attribute = AIE_TILE_TYPE_TILE << AIE_REGS_ATTR_TILE_TYPE_SHIFT,
 	 .soff = AIE_TILE_CORE_CLKCNTR_REGOFF,
 	 .eoff = AIE_TILE_CORE_CLKCNTR_REGOFF,
 	},
-- 
2.31.1

